xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -165,24 +167,12 @@ static BootstrapInfo parseConfig(String rawData) throws IOException {
     Map<String, CertificateProviderInfo> certProviders = null;
     if (certProvidersBlob != null) {
       certProviders = new HashMap<>(certProvidersBlob.size());
-      for (Map.Entry<String, ?> entry : certProvidersBlob.entrySet()) {
-        String name = entry.getKey();
-        Object value = entry.getValue();
-        if (!(value instanceof Map)) {
-          throw new IOException(
-              "Invalid bootstrap: invalid 'certificate_providers' entry for " + name);
-        }
-        @SuppressWarnings("unchecked")
-        Map<String, ?> valueMap = (Map<String, ?>) value;
-        Object pluginName = valueMap.get("plugin_name");
-        Object config = valueMap.get("config");
-        if (!(pluginName instanceof String) || !(config instanceof Map)) {
-          throw new IOException(
-              "Invalid bootstrap: invalid 'certificate_providers' entry for " + name);
-        }
-        @SuppressWarnings("unchecked")
+      for (String name : certProvidersBlob.keySet()) {
+        Map<String, ?> valueMap = JsonUtil.getObject(certProvidersBlob, name);
+        String pluginName = JsonUtil.getString(valueMap, "plugin_name");
```

Comment:
Since this is the parsing logic, the input (the boostrap JSON file) can contain anything, it needs to be defensive. You should always check if the retrieved result exists, a proper exception should be thrown if not. All exceptions should be checked, runtime exceptions will be a surprise.

Reply:
But `JsonUtil.getString()`, `JsonUtil.getObject()` etc (which is being used in `parseConfig` and you suggested I use) throw `ClassCastException` which is a `RuntimeException` so how is this different? Presumably you are talking about checkNotNull later inside `new CertificateProviderInfo`. I can have checkNotNull or checkArgument with a better message if that works.

Reply:
```
String pluginName = JsonUtil.getString(valueMap, "plugin_name");
if (pluginName == null) {
  throw new IOException("a descriptive string");
}
```

This is the defensive programming pattern I am talking about. This catches the following JSON file content error:
```
    "gcp_id": {
        "config": { ...MeshCA config... }
    }
```

Reply:
In the following example:
```
"gcp_id": {
        "plugin_name": 123,
        "config": { ...MeshCA config... }
    }
```
You'll get a `ClassCastException` from `JsonUtil.getString(valueMap, "plugin_name")` and that's a `RuntimeException` and not an `IOException`. Do you think that one needs to be addressed as well?

Reply:
> `parseConfig` is full of such cases and they will all need to be fixed (or have a common catch clause at the end to create a wrapper IOException around ClassCastException).

Yes, I think that could be ideal and this makes `readBootstrap()` have a better API guarantee.

Reply:
> Yes, I think that could be ideal and this makes `readBootstrap()` have a better API guarantee.

I can do that in this PR if that's what is blocking this PR from your point of view. Or I can do that in a separate PR. Lmk

Reply:
Up to you, that's very picky for error handlings. Here, my main point is regarding the API usage. From an API's perspective `JsonUtil.getString` may give you `null`, so callers may want to handle that properly.

Reply:
But what I am saying is there is no qualitative difference between a missing value for `"plugin_name"` and a value of wrong type for `"plugin_name"`. Currently in both cases they generate a `RuntimeException` (`NullPointerException` and `ClassCastException`) which is consistent. If you are thinking of API guarantees all these should be fixed instead of just selectively doing it for the new code in this PR. Do you agree?

Reply:
With your current implementation, if fields like "plugin_name" is missing, the application will be panic with a NullPointException pointing that your code has a bug.

For those `ClassCastException` cases, we are still able to see from the exception that the JSON file is malformed.